resolver/validator.go

```diff
@@ -0,0 +1,90 @@
+package resolver
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/renproject/darknode/jsonrpc"
+	"github.com/renproject/darknode/tx"
+	"github.com/renproject/darknode/txengine/txenginebindings"
+	"github.com/renproject/id"
+	v0 "github.com/renproject/lightnode/compat/v0"
+)
+
+// The lightnode Validator checks requests and also casts in case of compat changes
+type LightnodeValidator struct {
+	bindings *txenginebindings.Bindings
+	pubkey   *id.PubKey
+	store    v0.CompatStore
+}
+
+func NewVerifier(bindings *txenginebindings.Bindings, pubkey *id.PubKey, store v0.CompatStore) *LightnodeValidator {
+	return &LightnodeValidator{
+		bindings: bindings,
+		pubkey:   pubkey,
+		store:    store,
+	}
+}
+
+// The validator usually checks if the params are in the correct shape for a given method
+// We override the checker for certain methods here to cast invalid v0 params into v1 versions
+func (validator *LightnodeValidator) ValidateRequest(ctx context.Context, r *http.Request, req jsonrpc.Request) (interface{}, jsonrpc.Response) {
+	switch req.Method {
+	case jsonrpc.MethodQueryTx:
+		// Check if the params deserializes to v1 queryTx
+		// to check if we need to do compat or not
+		var v1params jsonrpc.ParamsQueryTx
+		// This will throw an error if it's a v1 query because the base64 encoding is different
+		if err := json.Unmarshal(req.Params, &v1params); err == nil {
```

Comment:
During submission we can clearly determine the version of a tx because the parms have a different shape.

Reply:
We might have to go ahead with the extra Redis lookup. Wouldn't want users losing their transactions because of chance.

Reply:
Actually just ran through the logic here again - we already check redis for each request (in the resolver); and only treat it as a v0 tx if the mapping is found, otherwise treat it as a v1 request. So clashes should not matter, and the only failure case would be if we failed to persist the tx submission in redis.